### DISCUSSION TITLE
Working with $filter in a dynamic odata model

### DISCUSSION THREAD

**Discussion by Remo** [OData/AspNetCoreOData#1311] [Q&A] [answered]
I've got a use cases where a very dynamic OData setup might come in handy.
This example was a great help: https://github.com/OData/AspNetCoreOData/tree/main/sample/ODataDynamicModel

I'm now trying to understand how I can access `$filter`. I expected that I could use something similar to this https://github.com/OData/AspNetCoreOData/blob/main/sample/ODataDynamicModel/Controllers/HandleAllController.cs#L33, but I cannot find anything related to filtering.

Can someone please point me in the right direction? Neither google nor copilot seem to know anything about this.

**Comment by Remo** [comment, marked as answer]
I tried to add `ODataQueryOptions` as an argument in my get method, but when I do that, I only get an 404 error back.
After some tinkering, I found out, that I can build the query options object myself.

```
IEdmModel model = Request.GetModel();
ODataQueryContext queryContext = new ODataQueryContext(model, edmEntityType, path);
ODataQueryOptions queryOptions = new ODataQueryOptions(queryContext, Request);
```

This seems to give me access to the filter and other things like orderby.

**Reply by gathogojr** [reply]
Please note that we don't support the FilterBinder in non-Edm model behaviour. If your plan it to get hold of the query options object and apply the filter (or other query options) by yourself, your approach here should work.

**Reply by Remo** [reply]
Thanks! I'm adapting my plans a bit. I've got a lot of legacy stuff for which I wanted to generate OData services on the fly. Since the legacy stuff changes a lot, I didn't want to use source generators, but rather something that doesn't re-compilation. However, I realized that I would basically have to implement the whole ORM layer myself which is insane. Instead I'm using Roslyn to generate my entities on the fly and then add them to a DbContext as well as EDM model for OData. I'm not going to win a design award for my architecture, but it's highly dynamic and very easy to maintain as it basically updates itself.

**Reply by lorenzocas** [reply]
Hello,
i'm trying to adapt https://github.com/OData/AspNetCoreOData/tree/main/sample/ODataDynamicModel to use entity framework, but i'm not sure how to implement the ef querying part in the sample. I think i should change the Get() Methods in MyDataSource class, but i'm not sure how. Could you please give me some hints how to do this ?

Thanks
Best regards

**Reply by Remo** [reply]
@lorenzocas I think this will probably go a bit off-topic, might be better to start a new discussion for this. In short: The example you listed doesn't use entity framework. I'm not quite sure why you want this dynamic model project as a basis. If you have traditional entity framework models, you've got proper POCO classes, hence no need for dynamic models?

**Comment by lorenzocas** [comment]
Hello,
I created a new discussion here:

https://github.com/OData/AspNetCoreOData/discussions/1317

Thanks
